### DISCUSSION TITLE
Add option to disable getServerSideProps on client-side navigation

### DISCUSSION THREAD

**Discussion by ziimakc** [vercel/next.js#19611] [Help] [answered]
# Feature request

## Is your feature request related to a problem? Please describe.

1. Function `getServerSideProps` runs on every request and there is no way to use apollo [existing cache to load data from on client-side navigation.](https://github.com/vercel/next.js/discussions/11957#discussioncomment-48814) It's will do unnecessary request to server if there is data already in client cache.

2. We need SSR only on initial render for SEO, there is no need for it to run every time and block page rander, it's better to show loading or page skeleton.

## Describe the solution you'd like

Add option to disable getServerSideProps on client-side navigation

```
export async function getServerSideProps(context) {
  return {
    props: {},
    disableOnClientSideNavigation: true,
  }
}
```

## Describe alternatives you've considered

Add another function like `getServerSideProps` thats will not run on client navigation.

OR

Add export config from page:
```
export pageConfig = {
    disableOnClientSideNavigation: true,
}
```

OR

Add option to router:
`router.push('user/1', undefined, { disableGetServerSideProps : true })`

## Additional context

Then you can use same graphql query in component and getStaticProps

```
const UserQuery = gql`
  query User {
    getUser {
      id
    }
  }
`

export function User(): JSX.Element {
  const { data } = useQuery(UserQuery)

  return (
      <div>user</div>
  )
}

export const getStaticProps: GetStaticProps = async (ctx) => {
  const apolloClient = initializeApollo(null)

    await apolloClient.query({ query: UserQuery})

  return {
    props: {
      initialApolloState: apolloClient.cache.extract(),
    },
  }
}
```

**Comment by balazsorban44** [comment]
From the docs:
> `shallow` - Update the path of the current page without rerunning `getStaticProps`, `getServerSideProps` or `getInitialProps`. Defaults to `fals`


https://nextjs.org/docs/api-reference/next/link
https://nextjs.org/docs/api-reference/next/router#routerpush

**Comment by ziimakc** [comment]
@balazsorban44 are you sure it will render new component and not just change the link?

**Comment by balazsorban44** [comment]
I am not sure I understand what you mean. When you transition from a page to another (by `Link` or `router`) you will render a new (page) component, yes. Adding the `shallow` option to either the `Link` or `router.push`, it will disable the `getServerSideProps` call.

**Comment by ziimakc** [comment]
@balazsorban44 
From docs:

The URL will get updated to /?counter=10. and the page won't get replaced, only the state of the route is changed.

Shallow routing only works for same page URL changes. For example, let's assume we have another page called pages/about.js, and you run this:

`router.push('/?counter=10', '/about?counter=10', { shallow: true })`

Since that's a new page, it'll unload the current page, load the new one and wait for data fetching even though we asked to do shallow routing.

So it's not a sollution.

**Comment by balazsorban44** [comment]
I see. Sorry I could not help. 😕

**Comment by stefanbackor** [comment]
Take a look at `getInitialProps` https://nextjs.org/docs/api-reference/data-fetching/getInitialProps
It does exactly what you need.

**Comment by vaniyokk** [comment]
Any activity on this?
It's really annoying problem!
Very common plugins (next-i18next, next-translate) are already dropped support for `getInitialProps`. All new examples are using `getServerSideProps` as preffered solution and there is NO WAY to eliminate extra network call every time route changed?
Nobody uses client-side cache? Hmm... don't believe.
Waiting for experimental react feature is not a solution for now.

You really need to think about it. I don't understand why this issue takes so little attention from community...

**Reply by zcmk123** [reply]
> Any activity on this?
> It's really annoying problem!
> Very common plugins (next-i18next, next-translate) are already dropped support for `getInitialProps`. All new examples are using `getServerSideProps` as preffered solution and there is NO WAY to eliminate extra network call every time route changed?
> Nobody uses client-side cache? Hmm... don't believe.
> Waiting for experimental react feature is not a solution for now.
> 
> You really need to think about it. I don't understand why this issue takes so little attention from community...

same here, i'm using next-i18next, every client router change, next will send a new request to get i18 translation, no cache......is annoying

**Comment by k4mr4n** [comment]
@ZiiMakc I solved this issue by using this on every getServerSideProps for the time being.

```js
const isFirstServerCall = req?.url?.indexOf('/_next/data/') === -1
```

**Reply by longprao** [reply]
Thank you so much, I have been using react-query to have client side caching and this check is a life saver, now I can have cache when moving between Next Links.

**Reply by k4mr4n** [reply]
@longprao  If your concern is calling the server on route change on the client-side, then, you should not use getServerSideProps at all
you should use getInitialProps. the nextJs team is saying that they will deprecate this but they can't and they won't anytime soon since there is no other replacement for getInitialProps yet.

**Reply by balthazar** [reply]
I assume this only works for next <12?

**Reply by ziimakc** [reply]
> I assume this only works for next <12?

Idn, I've started to use remix, where is no such a problem.

**Reply by hoffnung8493** [reply]
> > I assume this only works for next <12?
> 
> Idn, I've started to use remix, where is no such a problem.

But isnt Remix's loader function very similar to getServerSideProps. Loader will fetch data again from the remix server on page revisits, just like getServerSideProps

**Reply by damirJa** [reply]
Isn't @hoffnung8493 is right about the Remix having the same issue?

**Reply by hoffnung8493** [reply]
@damirJa At first glance, it did have the same issue. But it turns out u can disable it with [shouldRevalidate](https://remix.run/docs/en/main/route/should-revalidate).

But tbh, I dont use this very often, cuz with Remix, I no longer manage global state in frontend. I dont use react-query, redux etc. So I my usecase for `revalidate` is a lot less now. With Remix, I just keep the state in the server. Check this official[ blog post](https://remix.run/blog/remix-data-flow) if u want to learn more about it.

**Reply by damirJa** [reply]
Thanks @hoffnung8493 for your advice, this is a really useful and interesting solution! I will definitely test it out!

**Reply by hoffnung8493** [reply]
@damirJa also u can check out this [feature request](https://github.com/remix-run/remix/discussions/4950) that I have made to the Remix team. It pretty much tackles the same problem and its added to their roadmap!

**Comment by emzet93** [comment]
Any updates on that? Like @vaniyokk said, using `getServerSideProps` kills all benefits of using local cache. Currently we use `react-query` for data fetching and the perfect scenario would be to have some possibility to call `getServerSideProps` (or something similar) only once, on initial page load, and not calling it on client side routes transition and let react-query do the work.

Currently I managed to achieve exactly what I wanted by using `getInitialProps` on every page and checking inside if it's called on server or on the client. It looks something like that:
```
Home.getInitialProps = async () => {
   if (type of window !== "undefined") {
      return {};
   }
   
   const queryClient = new QueryClient();
   await queryClient.prefetchQuery(QueryKeys.homeData, fetchHomeData);

   return {
      dehydratedState: dehydrate(queryClient),
   };
}
```

It works exactly like I wanted to, but the thing is that `next js` docs does not recommend using `getInitialProps` and it says that it will be probably deprecated in the future. Also some libraries (e.g. `next-i18next`) are not compatible whit this method anymore.

It would be great if we could achieve the same results using well supported methods and not looking for workarounds. Especially considering how great and popular `react-query` became.

**Reply by k4mr4n** [reply]
I ended up with this snippet on `_app.tsx` to implement it once and to prevent repeating myself.

```tsx
NextApp.getInitialProps = async ({ Component, ctx, router }) => {
  try {
    let pageProps: any = {}
    const isServer = typeof window === 'undefined'
    let appProps
    // if the page is standAlone then it does not need appsProps
    if (Component.standAlone !== true) {
      appProps = await getAppInitialProps(ctx, Component.options, router)
    }
    if (Component.getInitialProps) {
      pageProps = await Component.getInitialProps(ctx, { isServer }) // here you can fetch the page initial props
    }
    const apolloState = ctx?.apolloClient?.cache.extract() // or react-query
    return { pageProps: { ...pageProps, ...appProps, apolloState } }
  } catch (error) {
    return { pageProps: { error } }
  }
}
```

**Reply by emzet93** [reply]
Nice, thanks!

I added a function wrapper for `getInitialProps` on the page level:

```
// wrapper
export const getPrefetchQueryWrapper =
  (action: () => Promise<unknown>) => async () => {
    if (isClientEnv()) {
      return {};
    }

    return action();
  };

// usage in page
HomePage.getInitialProps = getPrefetchQueryWrapper(async () => {
  const queryClient = new QueryClient();
  await queryClient.prefetchQuery(QueryKeys.homeData, fetchHomeData);

  return {
    dehydratedState: dehydrate(queryClient),
  };
});
```

But still it looks like a workaround and I'm just curios what is `next.js` team opinion on such cases 🤔

**Reply by mopsyshin** [reply]
And if you want to use `prefetchInfiniteQuery` in getServerSideProps there is no way caching your previous list data after clientside routing. In this situation, you can't use client caching but also restoring scroll position. You have to refetch all of data from top of the page.

So.... I'm using getInitialProps. there is no other option.

**Reply by ynifamily3** [reply]
If you use getInitialProps, Server-side code is exposed and delivered to the client bundle. The desired purpose can be achieved, but there may be risks such as security. Although highly sensitive information such as API token keys can be prevented from being exposed using environment variables (without NEXT_PUBLIC_ prefix). However, it should be taken into account that the server code is bundled and left on the client side.

**Comment by jaworek** [comment]
In the project that I currently work on with @satya164 we use such approach:
https://twitter.com/satya164/status/1542095637432438785

```
Page.getInitialProps = () => {
  if (__IS_SERVER__) {
    // Fetch data
  }
}
```

```
const nextConfig = {
  webpack: (config, {dev, webpack, isServer}) => {
    config.plugins.push(
      new webpack.DefinePlugin({
        __DEV__: dev,
        __IS_SERVER__: isServer,
      })
    );
    
    return config;
  }
}
```

We inject __IS_SERVER__ in the webpack config and use getInitialProps instead of getServerSideProps, so that the call isn't made on client-side navigation. The __IS_SERVER__ check avoids running the prefetch code on the client side.

Not too bad since at least it's readable and does exactly what we want. Hope there was a built-in way though.

**Comment by dgrcode** [comment]
I came across this discussion after asking something similar in StackOverflow ([link](https://stackoverflow.com/questions/73303466/skip-ssr-in-nextjs-when-i-already-have-the-data-cached-and-not-stale-in-the-clie)).

I'm also in the group of people using ReactQuery as client cache. I was hoping there was a way to skip SSR if the data collected during `getServerSideProps` was already in the client cache.

After reading this, two possibilities came to mind. I'm thinking out loud here and haven't tested any method, but sharing just to see what people think and possible points of failure.

## 1. Setting a cookie to skip parts of SSR
I'm not sure if this could lead to hydration issues, since the SSR page won't have data, but the client will use the data in the cache. It will depend on when the two pages (SSR and Client) are compared, but I don't know that.
```ts
export const getServerSideProps = (req, res) => {
  let data
  if (!context.req.cookies.skipSsr) {
    // do expensive fetching here
  }

  return {
    props: { data }
  }
}
```

## 2. Using two different components
You'll have a default component that uses the client cache, and a second component that fetches data during SSR.

The main issue here is the routes, since (AFAIK) you can't use two different pages for the same route. The proposal here is that the SSR page just have an ending `_`. That page does the fetching on the server, renders the page with the whole data, and sends it to the client. Then, on the client, the data is stored in the cache, and a redirect (replacing) is done to the standard url, without the `_`.
```ts
// page using SSR
// /pages/foo_

export const getServerSideProps = (req, res) => {
  const queryClient = new QueryClient()
  await queryClient.prefetchQuery( /* do your thing */ )

  return {
    props: { dehydratedState: dehydrate(queryClient) },
  }
}

const FooSsr = () => (
  const { data } = useQuery(['your-key'], getStuff, {
    refetchOnMount: false,
    refetchOnWindowFocus: false,
  })

  const router = useRouter()
  useEffect(() => {
    router.replace('/foo')
  }, [])

  return <FooPage data={data} />
)
```

```ts
// page client cache
// /pages/foo

const Foo = () => (
  const { data } = useQuery(['your-key'], getStuff, {
    refetchOnMount: false,
    refetchOnWindowFocus: false,
  })
  return <FooPage data={data} />
)
```
---
Thoughts on these approaches?

**Reply by mmmulani** [reply]
not the biggest fan of approach 1 as it'll incur a request to the backend unnecessarily, though approach 2 is more interesting. However, I'm not sure if it'll play well with Next.js's route analysis that happens at build time.

**Comment by livyk** [comment]
Any updates here? It seems like big issue for every application which uses react-query / appolo client etc

**Reply by stijnwesterhof** [reply]
The new Next.js router that is gonna be release soon (I hope) probably has the perfect solution for this problem. See [this](https://twitter.com/tkdodo/status/1558752788393476096) twitter discussion.

**Comment by IshmamR** [comment]
use `getInitialProps` instead of `getServerSideProps`. The page will get created only once while you are navigating using Link/router.
You could also make use of `shallow: true` option on route changes.

**Comment by dreamer01** [comment]
Found a way to use caching at `getServerSideProps` , [here](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#caching-with-server-side-rendering-ssr) 

But this will require for our deployment service to support caching, and won't help where we are using `react-query` for caching solution.

Waiting for `nested-routes` RFC to release for a better solution.

```js
export async function getServerSideProps({ req, res }) {
  res.setHeader(
    'Cache-Control',
    'public, s-maxage=10, stale-while-revalidate=59'
  )

  return {
    props: {},
  }
}
```

**Comment by mdegrees** [comment]
@leerob does Next13 React Server Component solve this issue?

**Comment by leerob** [comment, marked as answer]
For folks exploring the App Router: https://nextjs.org/docs/app/building-your-application/caching

**Reply by damirJa** [reply]
Thanks for the link, but unfortunately it doesn't apply to `page router` and is not an answer to the original question/feature request
